{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TRY NER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import spacy, re , json\n",
    "from spacy.matcher import Matcher\n",
    "from spacy.tokens import Span\n",
    "from collections import defaultdict\n",
    "from glob import glob\n",
    "\n",
    "class MacherBase(object):\n",
    "    def __init__(self, nlp, patterns, label):\n",
    "\n",
    "        self.matcher = Matcher(nlp.vocab)\n",
    "        self.matcher.add(label, None, *patterns)\n",
    "    def __call__(self, doc):\n",
    "        matches = self.matcher(doc)\n",
    "        seen_tokens = set()\n",
    "        new_entities = []\n",
    "        entities = doc.ents\n",
    "        start_end_list = []\n",
    "        for match_id,start,end in matches:\n",
    "            start_end_list.append([start,end])\n",
    "\n",
    "        start_end_length = len(start_end_list)\n",
    "        del_mark = []\n",
    "\n",
    "        for i in range(start_end_length-1):\n",
    "            i_start,i_end = start_end_list[i]\n",
    "\n",
    "            for j in range(i+1,start_end_length):\n",
    "                j_start,j_end = start_end_list[j]\n",
    "                \n",
    "                if j_start>i_end:\n",
    "                    break\n",
    "                if j_end> i_end:\n",
    "                    del_mark.append(i)\n",
    "\n",
    "        for m_index,match in enumerate(matches):\n",
    "            if m_index in del_mark:\n",
    "                continue\n",
    "            match_id, start, end = match\n",
    "            if start not in seen_tokens and end - 1 not in seen_tokens:\n",
    "                new_entities.append(Span(doc, start, end, label=match_id))\n",
    "                entities = [\n",
    "                    e for e in entities if not (e.start < end and e.end > start)\n",
    "                ]\n",
    "                seen_tokens.update(range(start, end))\n",
    "\n",
    "        doc.ents = tuple(entities) + tuple(new_entities)\n",
    "        return doc\n",
    "\n",
    "        \n",
    "class TiterMatcher(MacherBase):\n",
    "    name = \"Titer_matcher\"\n",
    "    \n",
    "    def __init__(self, nlp, patterns, label):\n",
    "        MacherBase.__init__(self, nlp, patterns, label)\n",
    "\n",
    "class YieldMatcher(MacherBase):\n",
    "    name = \"Yield_matcher\"\n",
    "\n",
    "    def __init__(self, nlp, patterns, label):\n",
    "        MacherBase.__init__(self, nlp, patterns, label)\n",
    "        \n",
    "class ProdRateMatcher(MacherBase):\n",
    "    name = \"ProdRate_matcher\"\n",
    "\n",
    "    def __init__(self, nlp, patterns, label):\n",
    "        MacherBase.__init__(self, nlp, patterns, label)\n",
    "        \n",
    "\n",
    "def nlp_change():\n",
    "    nlp = spacy.load(\"en_core_web_sm\")\n",
    "\n",
    "                   \n",
    "    titer_patterns = [\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"^\\/[Ll]\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"^.?[Ll]\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+.?g$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"^[Ll]\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g·[Ll]\\-1$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"(^[0-9]+(\\.[0-9]{1,2})?$)\"}},{\"TEXT\":{\"REGEX\": \"(M|mol)$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"(^[0-9]+(\\.[0-9]{1,2})?$)\"}},{\"TEXT\":{\"REGEX\": \"(M|mol)$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"^[Ll]\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"^\\d+(.+)?(M|mol)$\"}}],\n",
    "                      [{\"POS\": \"NUM\"},{\"TEXT\":{\"REGEX\": \"\\+\"}},{\"TEXT\":{\"REGEX\": \"-\"}},{\"TEXT\":{\"REGEX\": \"^\\d+\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"^\\/?[Ll]\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"^\\d+m?g·?[Ll]\\(?\\-1$\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"^\\d+g·?[Ll]\\-1$\"}}]\n",
    "                     ]\n",
    "\n",
    "\n",
    "    titer_matcher = TiterMatcher(nlp, titer_patterns, 'TITER')\n",
    "    nlp.add_pipe(titer_matcher,before='ner')\n",
    "\n",
    "    # DCW: dry cell weight\n",
    "\n",
    "    yield_patterns = [\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"g$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\/?-\"}},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"g$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"g\\-1$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+$\"}},{\"TEXT\":{\"REGEX\": \"[Ll]$\"}},{\"TEXT\":{\"REGEX\": \"\\/\"}},{\"TEXT\":{\"REGEX\": \"g$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"\\/?\"}},{\"TEXT\":{\"REGEX\": \"g‐glucose$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"\\/?\"}},{\"TEXT\":{\"REGEX\": \"gDC?W$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \"g\\-1$\"}},{\"TEXT\":{\"REGEX\": \"DC?W$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g$\"}},{\"TEXT\":{\"REGEX\": \".+\"}},{\"TEXT\":{\"REGEX\": \"per\"}},{\"TEXT\":{\"REGEX\": \"gram$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"TEXT\":{\"REGEX\": \".+\"}},{\"TEXT\":{\"REGEX\": \"per\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"TEXT\":{\"REGEX\": \"\\/?\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"TEXT\":{\"REGEX\": \"(mol-1|mol−1)$\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\",\"OP\": \"?\"},{\"ORTH\":\"(\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"C$\"}},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"TEXT\":{\"REGEX\": \"(mol-1|mol−1)$\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\",\"OP\": \"?\"},{\"ORTH\":\"(\",\"OP\": \"?\"},{\"ORTH\":\"C\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"ORTH\":\"C\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"(mol-1|mol−1)$\"}},{\"ORTH\":\")\",\"OP\": \"?\"}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\"},{\"ORTH\":\"molar\"},{\"TEXT\":{\"REGEX\": \"^yield\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\"},{\"ORTH\":\"theoretical\"},{\"TEXT\":{\"REGEX\": \"^yield\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\"},{\"ORTH\":\"theoretical\"},{\"TEXT\":{\"REGEX\": \"^yield\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"[0-9]+(\\.[0-9]{1,2})?\"}},{\"ORTH\":\"%\"},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"ORTH\":\"/\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"mol$\"}}],\n",
    "                      [{\"TEXT\":{\"REGEX\": \"[0-9]+(\\.[0-9]{1,2})?.g$\"}},{\"TEXT\":{\"REGEX\": \"\\/$\"}},{\"TEXT\":{\"REGEX\": \"g$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"(^[0-9]+(\\.[0-9]{1,2})?$)\"}},{\"ORTH\":\"%\"},{\"TEXT\":{\"REGEX\": \"mol$\"}},{\"ORTH\":\"/\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"mol$\"}}],\n",
    "                      #[{\"TEXT\":{\"REGEX\": \"^yield\"}},{\"ORTH\":\"of\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"ORTH\":\"%\"}]\n",
    "                     ]\n",
    "\n",
    "    yield_matcher = YieldMatcher(nlp, yield_patterns, 'YIELDS')\n",
    "    nlp.add_pipe(yield_matcher,before='ner')\n",
    "\n",
    "    prodrate_patterns = [\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"(M|mol)$\"}},{\"TEXT\":{\"REGEX\": \"\\/?\"}},{\"TEXT\":{\"REGEX\": \"^(min|hr?|d|days?)$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g·[Ll].+·(min|hr?|d|days?)$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g?\"}},{\"TEXT\":{\"REGEX\": \"\\/?\"}},{\"TEXT\":{\"REGEX\": \"^[Ll]\"}},{\"ORTH\":\"/\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"^(min|hr?|d|days?)$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g?\"}},{\"ORTH\":\"/\"},{\"TEXT\":{\"REGEX\": \"^([Ll]·)?(min|hr?|d|days?)$\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g?\"}},{\"ORTH\":\"/\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"^(min|hr?|d|days?)\"}},{\"ORTH\":\"/\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"^[Ll]\"}}],\n",
    "                      [{\"POS\": \"NUM\",\"OP\": \"?\"},{\"ORTH\":\"+\",\"OP\": \"?\"},{\"ORTH\":\"-\",\"OP\": \"?\"},{\"TEXT\":{\"REGEX\": \"\\d+\"}},{\"TEXT\":{\"REGEX\": \"g?\"}},{\"TEXT\":{\"REGEX\": \"^[Ll]-1\"}},{\"TEXT\":{\"REGEX\": \"^(min|hr?|d|days?)\"}}]                    \n",
    "                     ]\n",
    "    prodrate_matcher = ProdRateMatcher(nlp, prodrate_patterns, 'PRODRATE')\n",
    "    nlp.add_pipe(prodrate_matcher,before='ner')\n",
    " \n",
    "    nlp.remove_pipe('ner')\n",
    "    return nlp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['201.52 mg/L**#**TITER**#**161**#**172']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def prod_token(text):\n",
    "    doc = nlp(text)\n",
    "    \"\"\"\n",
    "    for tok in doc:\n",
    "        print(tok.text,\"-->\", tok.pos_,\"-->\",tok.ent_type_,\"-->\",tok.i)\n",
    "    \"\"\"\n",
    "    prod_record = []\n",
    "    for ent in doc.ents:\n",
    "        #print(dir(ent))\n",
    "        ent_start = str(int(ent.start_char)+titel_length)\n",
    "        ent_end = str(int(ent.end_char)+titel_length)\n",
    "        #print(ent.text, ent.start_char, ent.end_char, ent.label_,ent.start,ent.end)\n",
    "\n",
    "        sub_prod = '**#**'.join([ent.text,ent.label_,ent_start,ent_end])\n",
    "        prod_record.append(sub_prod)\n",
    "    return prod_record\n",
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "nlp = nlp_change()\n",
    "text = 'Then, over-expression of 2,3-DHBA synthase and 2,3-DHBA 1-monoxygenase achieved synthesis of pyrogallol in Escherichia coli, with a titer of 201.52 mg/L at 24 h.'\n",
    "titel_length =20\n",
    "#pubmed_id = '20000'\n",
    "prod_token(text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chem NER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MK-7', 'menaquinone-7']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from chemdataEx.doc.document import Document\n",
    "\n",
    "def chemdataEx_chem(abstract_text):\n",
    "    doc = Document(abstract_text)\n",
    "    doc_chems = doc.cems\n",
    "    ChemsText = []\n",
    "    for chems in doc_chems:\n",
    "        chem_text = chems.text\n",
    "        chem_start= chems.start\n",
    "        chem_end = chems.end        \n",
    "        one_chem_digs = [c_dig for c_dig in range(chem_start,chem_end)]\n",
    "        ChemsText.append(chem_text) \n",
    "    ChemsText = list(set(ChemsText))  \n",
    "    return ChemsText\n",
    "abstract_text = 'The QS system can dynamically balance the relationship between efficient synthesis of the target product and cell growth. Finally, we validated the usefulness of this strategy by dynamic control of menaquinone-7 (MK-7 ) synthesis in Bacillus subtilis 168.'\n",
    "ChemsList = chemdataEx_chem(abstract_text)\n",
    "ChemsList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chems_list: ['2,3-DHBA']\n",
      "2,3-DHBA 25 33 CHEM 6 7\n",
      "2,3-DHBA 47 55 CHEM 9 10\n",
      "201.52 mg/L 141 152 TITER 23 27\n"
     ]
    }
   ],
   "source": [
    "from spacy.matcher import PhraseMatcher\n",
    "class ChemsMatcher(object):\n",
    "    # detected chem is longer one\n",
    "    name = \"Chems_matcher\"\n",
    "\n",
    "    def __init__(self, nlp, terms, label):\n",
    "        patterns = [nlp.make_doc(text) for text in terms]\n",
    "        self.matcher = PhraseMatcher(nlp.vocab)\n",
    "        self.matcher.add(label, None, *patterns)\n",
    "\n",
    "    def __call__(self, doc):\n",
    "        matches = self.matcher(doc)\n",
    "        seen_tokens = set()\n",
    "        new_entities = []\n",
    "        entities = doc.ents\n",
    "        start_end_list = []\n",
    "        for match_id,start,end in matches:\n",
    "            start_end_list.append([start,end])\n",
    "\n",
    "        start_end_length = len(start_end_list)\n",
    "        del_mark = []\n",
    "\n",
    "        for i in range(start_end_length-1):\n",
    "            i_start,i_end = start_end_list[i]\n",
    "\n",
    "            for j in range(i+1,start_end_length):\n",
    "                j_start,j_end = start_end_list[j]\n",
    "                \n",
    "                if j_start>i_end:\n",
    "                    break\n",
    "                if j_end> i_end:\n",
    "                    del_mark.append(i)\n",
    "        \n",
    "        #print(del_mark)\n",
    "        for m_index,match in enumerate(matches):\n",
    "            if m_index in del_mark:\n",
    "                continue\n",
    "            match_id, start, end = match\n",
    "            if start not in seen_tokens and end - 1 not in seen_tokens:\n",
    "                new_entities.append(Span(doc, start, end, label=match_id))\n",
    "                entities = [\n",
    "                    e for e in entities if not (e.start < end and e.end > start)\n",
    "                ]\n",
    "                seen_tokens.update(range(start, end))\n",
    "\n",
    "        doc.ents = tuple(entities) + tuple(new_entities)\n",
    "        return doc \n",
    "#\"\"\"\n",
    "nlp = spacy.load(\"en_core_web_sm\")    \n",
    "nlp = nlp_change()\n",
    "#\"\"\"\n",
    "chems_list = ['2,3-DHBA']\n",
    "print(\"chems_list:\",chems_list)\n",
    "chems_matcher = ChemsMatcher(nlp,chems_list,\"CHEM\")\n",
    "nlp.add_pipe(chems_matcher,before='Titer_matcher')\n",
    "text = 'Then, over-expression of 2,3-DHBA synthase and 2,3-DHBA 1-monoxygenase achieved synthesis of pyrogallol in Escherichia coli, with a titer of 201.52 mg/L at 24 h.'\n",
    "doc = nlp(text)\n",
    "for ent in doc.ents:\n",
    "    #print(dir(ent))\n",
    "    print(ent.text, ent.start_char, ent.end_char, ent.label_,ent.start,ent.end)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocessing the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 ../prodcurator/3955571_curator.json\n",
      "As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.\n",
      "1 ../prodcurator/12932830_curator.json\n",
      "The oligonucleosomes assembled from 12-repeat DNA and saturating amounts of core histone octamer plus histone H5 are compacted, in the presence of 1 mM free magnesium ions, to the level of the 30-nm fiber.\n",
      "2 ../prodcurator/16507371_curator.json\n",
      "3 ../prodcurator/27372134_curator.json\n",
      "Dissolved oxygen (DO) of UASS and AF reactors kept around 1.39+-0.27 and 0.99+-0.38mg/L, respectively.\n",
      "4 ../prodcurator/28311_curator.json\n",
      "It is, however, markedly labilized by an increase in the ionic strength of the medium brought about by the addition of 0.2 M potassium chloride or in pH above 9.\n",
      "5 ../prodcurator/2877616_curator.json\n",
      "Ornithine and arginine (5 to 20 mM), but not glutamic acid or proline, exerted a concentration-dependent stimulatory effect on the biosynthesis of clavulanic acid in both resting-cell cultures and long-term fermentations of Streptomyces clavuligerus.\n",
      "Both arginase activity and clavulanic acid formation were enhanced simultaneously by supplementing the production medium with 10 mM arginine.\n",
      "6 ../prodcurator/15095900_curator.json\n",
      "For these reasons, the effects of atrazine (5 and 50 microg/L) on several endpoints related to reproductive fitness were examined in fathead minnows (Pimephales promelas) in a 21-d static exposure.\n",
      "Estradiol (0.5 microg/L) was included as a positive-control treatment.\n",
      "7 ../prodcurator/28455805_curator.json\n",
      "8 ../prodcurator/16788051_curator.json\n",
      "9 ../prodcurator/350285_curator.json\n",
      "10 ../prodcurator/16707670_curator.json\n",
      "Analysis of the ectoine and hydroxyectoine contents of an ectABC ectD mutant strain fed with 1 mM ectoine or hydroxyectoine demonstrated that ectD is required for the main ectoine hydroxylase activity in C. salexigens.\n",
      "Although in minimal medium at 37 degrees C the wild-type strain grew with 0.5 to 3.0 M NaCl, with optimal growth at 1.5 M NaCl, at 45 degrees C it could not cope with the lowest (0.75 M NaCl) or the highest (3.0 M NaCl) salinity, and it grew optimally at 2.5 M NaCl.\n",
      "The ectD mutation caused a growth defect at 45 degrees C in minimal medium with 1.5 to 2.5 M NaCl, but it did not affect growth at 37 degrees C at any salinity tested.\n",
      "With 2.5 M NaCl, the ectD mutant synthesized 38% (at 37 degrees C) and 15% (at 45 degrees C) of the hydroxyectoine produced by the wild-type strain.\n",
      "11 ../prodcurator/19697159_curator.json\n",
      "The highest hydrogen-production rate (i.e., 0.18 mL/mg/d or 7.3 micromol/mg/d) was observed in cultures with an initial medium nitrate concentration of 1 mM at a light intensity of 100 micromol/m(2) s. The addition of photosystem II (PSII) inhibitor 3-(3,4-dichlorophenyl)-1,1-dimethylurea (DCMU) did not reduce hydrogen-production rates relative to unchallenged controls for 50 to 150 h, and intracellular glycogen concentrations decreased  during the hydrogen generation period.\n",
      "12 ../prodcurator/16100389_curator.json\n",
      "13 ../prodcurator/16347558_curator.json\n",
      "F(420) concentration in M. bryantii ranged from 1.84 to 3.65 mumol/g of protein; and in M. barkeri grown with methanol it ranged from 0.84 to 1.54 mumol .\n",
      "14 ../prodcurator/24241880_curator.json\n",
      "Isolated protoplasts required 50 mM Ca(2+) in the washing solution and 100 mM Ca(2+) in the culture medium to prevent lysis.\n",
      "15 ../prodcurator/11108281_curator.json\n",
      "The addition of hormones [36 nM estradiol-17beta, 1 microM medroxyprogesterone acetate, and 100 ng/ml relaxin] to IL-1beta-treated cells induced insulin-like growth factor binding protein-1 (IGFBP-1) messenger RNA expression after 3/ds of incubation.\n",
      "16 ../prodcurator/19838456_curator.json\n",
      "Pretreatment of Escherichia coli cultures with the iron chelator 2,2'-dipyridyl (1 mM) protects against the lethal effects of low concentrations of hydrogen peroxide (<15 mM).\n",
      "However, at H(2)O(2) concentrations equal to or greater than 15 mM, dipyridyl pretreatment increases lethality and mutagenesis, which is attributed to the formation of different types of DNA lesions.\n",
      "We show here that pretreatment with dipyridyl (1 mM) prior to challenge with high H(2)O(2) concentrations (>or=15 mM) induced mainly G:C-->A: We also show that in the fpg nth double mutant, the rpoB gene mutation (RifS-RifR) induced by 20 mM H(2)O(2) alone (20X higher) was increased in 20 mM H(2)O(2) and dipyridyl-treated cultures (110X higher), suggesting additional and/or different lesions in cells treated with H(2)O(2) under iron deprivation.\n",
      "17 ../prodcurator/20450489_curator.json\n",
      "18 ../prodcurator/20093013_curator.json\n",
      "Inocula from diverse locations were tested for their ability to reduce nitrate and perchlorate in synthetic ion exchange spent brine (45 g/l NaCl) using H(2)-based MBfRs.\n",
      "19 ../prodcurator/10873545_curator.json\n",
      "When the cultivation temperature of recombinant E. coli strains was lowered to 25 degrees C and the isopropyl-beta-d-thiogalactopyranoside (IPTG) concentration used for induction was decreased to as low as 0.02 mM, a total of about 33 mg of recombinant GDE can be isolated from a liter culture as estimated by amylo-1,6-glucosidase activity.\n",
      "20 ../prodcurator/21635572_curator.json\n",
      "21 ../prodcurator/10872208_curator.json\n",
      "22 ../prodcurator/6327521_curator.json\n",
      "23 ../prodcurator/24215455_curator.json\n",
      "The ligand-bound aptamer structure is essentially the same in cells and in buffer at 1 mM Mg(2+), the approximate Mg(2+) concentration we measured in cells.\n",
      "24 ../prodcurator/23567708_curator.json\n",
      "A significant reduction in nitrogen, between 34.7 and 73.7 mg /L, and particularly in phosphorus concentration, between 15.4 and 15.9 mg /L, was observed in drainage solution collected from commercial greenhouse production.\n",
      "25 ../prodcurator/7759392_curator.json\n",
      "The PER-1 producers and their transconjugants were highly resistant to ceftazidime and aztreonam (MIC &gt; or = 128 mg/L) but not to carbapenems or latamoxet.\n",
      "Piperacillin insusceptibility was marginal (MIC 8 mg/L).\n",
      "Clavulanate 4 mg/L, but not tazobactam 4 mg/L, reversed resistance to ceftazidime and carbenicillin.\n",
      "This reflected the present assays being in 0.1 M phosphate buffer pH 7.0, whereas the previous were pH-stat-regulated; concentrated phosphate reduced enzyme activity against ceftazidime, but not against cephaloridine.\n",
      "26 ../prodcurator/18576004_curator.json\n",
      "The continuous cofermentation performance of xylose-fermenting Zymomonas mobilis at 30 degrees C and pH 5.5 was characterized using a pure-sugar feed solution that contained 8 g/L glucose and 40 g/L xylose.\n",
      "Whereas xylose and cell-mass concentrations increased gradually with increasing dilution rate, ethanol concentrations and ethanol yields on available sugars remained approximately constant at 20-22 g/L and 80-90% of theoretical, respectively.\n",
      "Volumetric and specific ethanol productivities increased linearly with increasing dilution rate, rising from approx 1.0 g/L/h or g/g/h) at a dilution rate of 0.04/h to approx 2.0 g/L/h or g/g/h) at a dilution rate of 0.10/h.\n",
      "Similarly, specific sugar-utilization rates increased from approx 2.0 g/g/h at dilution rate 0.04/h to approx 3.5 g/g/h at dilution rate of 0.10/h.\n",
      "27 ../prodcurator/22268814_curator.json\n",
      "A quadratic response model was constructed through RSM designs, leading to a 57 5% increase of the growth-associated biosurfactant production by Acinetobacter sp. YC-X 2 with an optimized medium: beef extract 3 12 g/L ; peptone 20 87 g/L ; NaCl 1 04 g/L; and n-hexadecane 1 86 g/L .\n",
      "28 ../prodcurator/26125122_curator.json\n",
      "The decrease of Fe supply from 100 muM NaFe(III)-EDTA to 0.5 muM NaFe(III)-EDTA reduced growth and photosynthetic pigments similarly in the six barley varieties.\n",
      "29 ../prodcurator/9298537_curator.json\n",
      "30 ../prodcurator/23093384_curator.json\n",
      "By reinforcing the direct butanol-forming flux in C. acetobutylicum, 18.9 g/L of butanol was produced, with a yield of 0.71 mol butanol/mol glucose by batch fermentation,  which are 160% and 245% higher than those obtained with the wild type.\n",
      "By fed-batch culture of this engineered strain with in situ recovery, 585.3 g of butanol was produced from 1,861.9 g of glucose, with the yield of 0.76 mol butanol/mol glucose and productivity of 1.32 g/L/h.\n",
      "31 ../prodcurator/15701400_curator.json\n",
      "32 ../prodcurator/1587284_curator.json\n",
      "33 ../prodcurator/15635147_curator.json\n",
      "34 ../prodcurator/2121953_curator.json\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The polylysine-synthesizing ability by staphylococcal ribosomes increased up to about two times as much as that by E. coli Q13 ribosomes, when S-100 from E. coli Q13 was mixed with staphylococcal ribosomes which had been washed with a high salt HEPES buffer containing 10 mM HEPES, 1 mM EGTA, 16 mM magnesium acetate, 1.0 M ammonium chloride and 0.1 mM dithiothreitol (pH7.6).\n",
      "35 ../prodcurator/18584754_curator.json\n",
      "By limiting phosphorous concentration, HBsAg expression level for the YNN27/p2micro-S11 strain with inducible PHO5 promoter reached 0.2-0.3 mg/L .\n",
      "By controlling nutrient addition rate and dissolved oxygen concentration, HBsAg concentrations of 3-10 mg/L were achieved in 60-70 h fermentation using the YNN27/pDCB-S2 strain with the constitutive GPD promoter.\n",
      "36 ../prodcurator/1735138_curator.json\n",
      "37 ../prodcurator/16299003_curator.json\n",
      "Exogenous application of 10 microM ABA leads to swelling, root hair formation and initiation of lateral root primodia in the tips of young, seminal rice roots.\n",
      "38 ../prodcurator/25064336_curator.json\n",
      "This process provided the liquid fraction containing 111.4 g /L of sugars, corresponding to 681.0mM as monomeric sugars, from the original liquid fraction (181.1mM monomeric sugars).\n",
      "Xylose-fermenting recombinant Saccharomyces cerevisiae produced 34.5 +- 2.2 g /L ethanol from the 0.8 times liquid fraction (76% of theoretical yield).\n",
      "39 ../prodcurator/9404012_curator.json\n",
      "40 ../prodcurator/15018106_curator.json\n",
      "We performed a study which indicates that, as opposed to what happens in bacteria, the activity of NADP-GDH is fivefold higher when the halotolerant yeast Debaryomyces hansenii is grown in the presence of 1 M NaCl, compared with growth in media with no added salt.\n",
      "41 ../prodcurator/25456058_curator.json\n",
      "In PdCl2 supplemented experiments, the hydrogen yields of E. cloacae and mixed culture were 1.39 +- 0.07 and 2.11 +- 0.11 mol H2/mol glucose, respectively, with 5.0 mg/L of PdCl2.\n",
      "However, the highest hydrogen yields of E. cloacae and mixed culture were 1.48 +- 0.04 and 2.48 +- 0.09 mol H2/mol glucose, respectively at 5.0 mg/L of PdNPs supplementation.\n",
      "42 ../prodcurator/1464842_curator.json\n",
      "L-Arginine (10 mM), but not D-arginine (10 mM), prevented the inhibition by L-NMMA .\n",
      "Neither tetrodotoxin (TTX, 1 microM), nor the cyclo-oxygenase inhibitor, indomethacin (10 microM), altered the muscarinic vasodilatation or the inhibitory effect of L-NMMA .\n",
      "This vasodilatation was unaffected by L-NMMA but was abolished by the guanylate cyclase inhibitor, methylene blue (10 microM).\n",
      "The vasodilatation produced by ganglionic stimulation (10 Hz, 10 s) was blocked by TTX and the muscarinic receptor antagonist, 4-diphenylacetoxy-N-methyl-piperidine methiodide (4-DAMP, 1 microM).\n",
      "43 ../prodcurator/26089439_curator.json\n",
      "44 ../prodcurator/30338986_curator.json\n",
      "The sensor detected hexanoic, heptanoic and octanoic acid over a linear range up to 2, 1.5, and 0.75 mM, respectively, but did not show a linear response to decanoic and dodecanoic acid.\n",
      "45 ../prodcurator/10683164_curator.json\n",
      "In both animals, the SHb fraction increased with increasing Hb and sulphide concentrations up to 100 micromol/L and 1 mmol/L, respectively, and was favoured by a low pH .\n",
      "Within trout red blood cells, approximately 7 % of total haemoglobin was converted to SHb during 60 min of incubation (with 2.5 mmol/L sulphide), inducing a 20 % loss of haemoglobin oxygen-saturation.\n",
      "46 ../prodcurator/22991975_curator.json\n",
      "Isolates displaying minimum inhibitory concentrations (MICs) of > 2 mg/l by Etest to cefpodoxime and/or MICs > 1 mg/l to any other ESCs were included (n = 54; 35 E. coli, 11 K. pneumoniae, and 8 K. oxytoca).\n",
      "47 ../prodcurator/8096_curator.json\n",
      "48 ../prodcurator/21950118_curator.json\n",
      "Maximum indicators of SAS synthesis (concentration of extra cellular SAS is 7.2 g/l; factor of emulsification of the cultural liquid 50%; SAS yield from the substrate 50%) have been observed at 60-70% concentration of dissolved oxygen from the saturation level with aerial oxygen (pH 8.0) fractional supply of the substrate by portions each being 0.3-0.4% every 5-6 h to a final volume concentration of 2.4% and with the use of 10% inoculate grown until mid-exponential phase on the medium with 1.0 vol % of n-hexadecane.\n",
      "49 ../prodcurator/26930796_curator.json\n",
      "50 ../prodcurator/1429441_curator.json\n",
      "The deleterious effects of low pH could be partially prevented by 1 M sorbitol in the suspension of cells lacking sphingolipid.\n",
      "51 ../prodcurator/8226535_curator.json\n",
      "52 ../prodcurator/17558781_curator.json\n",
      "The performance of a bench-scale submerged membrane bioreactor (MBR) equipped with ultrafiltration membranes (ZENON) was investigated at different mixed liquor suspended solid (MLSS) concentrations (3069, 4314 and 6204 mg/L).\n",
      "53 ../prodcurator/25015887_curator.json\n",
      "The activities of mc(2)-155(pSmo) cells were 0.94, 0.57, 0.12, and 0.04 nmol/min/mg of protein with ethene, ethane, propane, and butane as substrates, respectively.\n",
      "54 ../prodcurator/17892288_curator.json\n",
      "55 ../prodcurator/17968511_curator.json\n",
      "When exposed to 200 mM NaCl, the dual gene-transformed plants displayed greater accumulation of betaine and Na(+) than their wild-type counterparts.\n",
      "56 ../prodcurator/2524572_curator.json\n",
      "57 ../prodcurator/3075658_curator.json\n",
      "Under aerobic growth conditions with glucose the acetyl-CoA and malonyl-CoA concentrations varied over a range of 0.05-1.5 nmol (mg dry wt)-1 (20-600 microM) and 0.01-0.23 nmol (mg dry wt)-1 (4-90 microM), respectively.\n",
      "58 ../prodcurator/7972519_curator.json\n",
      "59 ../prodcurator/28191902_curator.json\n",
      "We apply our module to identify the optimal point to redirect glycolytic flux into heterologous engineered pathways in Escherichia coli, resulting in titers of myo-inositol increased 5.5-fold and titers of glucaric acid increased from unmeasurable to >0.8 g/L, compared to the parent strains lacking dynamic flux control.\n",
      "We also used our module to control flux into aromatic amino acid biosynthesis to increase titers of shikimate in E. coli from unmeasurable to >100 mg/L.\n",
      "60 ../prodcurator/17651442_curator.json\n",
      "The reaction was stimulated by fumarate (> 500 microM), implicating the urea cycle enzyme argininosuccinate lyase (EC 4.3.2.1), which reversibly converts arginine and fumarate to argininosuccinate.\n",
      "61 ../prodcurator/17508726_curator.json\n",
      "62 ../prodcurator/12443989_curator.json\n",
      "IBMX (100 microM), an inhibitor of PDE, whose effects mimic those of low Ca(2+) on phototransduction, increased responses of both rod-ON-BCs and cone-ON-BCs, suggesting that the distinct effects of low Ca(2+) described above are attributable to differential modulation of signal transfer from different types of photoreceptors to BCs.\n",
      "Moreover, scotopic ERG P III responses, reflecting the rod activity, were potentiated both in low Ca(2+) and in the presence of IBMX (100 microM).\n",
      "63 ../prodcurator/6425262_curator.json\n",
      "Carbon monoxide dehydrogenase activity was inhibited 85% by 10 microM KCN and was rapidly inactivated by O2.\n",
      "Analysis of enzyme kinetic properties revealed an apparent Km of 5 mM for CO and a Vmax of 1,300 U/mg of protein.\n",
      "64 ../prodcurator/7959037_curator.json\n",
      "Transcription of the bvg-regulated promoters is repressed by modulating environmental factors such as 50 mM MgSO4, 10 mM nicotinic acid (NA) or low temperature (25 degrees C).\n",
      "We have isolated a spontaneous mutant (SK170) which expresses virulence genes at either 25 degrees C, or in the presence of 1-5 mM NA, or 10-50 mM MgSO4.\n",
      "Virulence factors in strain SK170 are still repressed by higher concentrations of NA (10 mM), or by a combination of low temperature (25 degrees C) and one of the other modulating agents.\n",
      "65 ../prodcurator/11587440_curator.json\n",
      "Glucose, acetate and fructose were tested as growth substrates, and the highest demeton-S-methyl biotransformation average rate (0.78 mg/L/h) and maximum instantaneous rate (1.4 mg/L/h) were achieved on fructose.\n",
      "66 ../prodcurator/11344454_curator.json\n",
      "The optimum temperature was around 30 degrees C, while the most suitable substrate concentrations were 100 mM each for glycerol and benzoic anhydride, respectively.\n",
      "However, when excessive anhydride (e.g., 200 mM) was used, the produced MBG could be further transformed into 1,3-dibenzoyl glycerol (DBG) by the same enzyme with a priority to (S)-MBG, resulting in a significant improvement of the product optical purity from ca. 50-70% e.e.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Under optimal conditions (100 mM glycerol, 100-200 mM benzoic anhydride, dioxane, 25-30 degrees C), the enzymatic synthesis of (R)-MBG was successfully operated in a packed-bed reactor for about 1 week, with an average productivity of 0.79 g MBG/day/g biocatalyst in the case of continuous operation and 0.94 g MBG/day/g biocatalyst in the case of semicontinuous operation.\n",
      "67 ../prodcurator/24068697_curator.json\n",
      "68 ../prodcurator/22133602_curator.json\n",
      "The finally genetic strain YJM13 harboring the MVA pathway and ispS(Pa) gene could accumulate isoprene up to 2.48 mg/l and 532 mg/l under the flask and fed-batch fermentation conditions, respectively, which is about three times and five times to the control strain.\n",
      "69 ../prodcurator/24376888_curator.json\n",
      "70 ../prodcurator/28336295_curator.json\n",
      "Biomass productivity was highest for AC768 strain with 1.8 g/L/d, while hydrocarbon production ranged from none to up to 42% per gram biomass dry weight, with Showa showing the highest hydrocarbon content followed by AC761.\n",
      "Glucose and galactose are the main monosaccharides in most strains and fucose content reached 463mgL-1 in CCALA778.\n",
      "71 ../prodcurator/983117_curator.json\n",
      "72 ../prodcurator/18597343_curator.json\n",
      "Improvement of the molar yield (moles acid produced per moles glucose consumed) of L-malic acid was obtained mainly by increasing the agitation rate (to 350 rpm) and the Fe(z+) ion concentration (to 12 mg/L) and by lowering the nitrogen (to 271 mg/L) and phosphate concentrations (to 1.5 mM) in the medium.\n",
      "The fermentation conditions developed using the above mentioned factors and 9% CaCO(3) in the medium resulted in a high concentration (113 g/L L-malic acid from 120 g/L glucose utilized) and a high overall productivity (0.59 g/L/h) of L-malic acid.\n",
      "73 ../prodcurator/20154256_curator.json\n",
      "74 ../prodcurator/24290677_curator.json\n",
      "75 ../prodcurator/9255947_curator.json\n",
      "This improved labeling scheme increases the sensitivity of the relaxation experiments by approximately fivefold when compared to randomly fractionally 13C-labeled protein, allowing high-quality measurements on relatively dilute (&lt; 1 mM) protein samples at a relatively low cost.\n",
      "76 ../prodcurator/7658603_curator.json\n",
      "77 ../prodcurator/21370229_curator.json\n",
      "At glucose concentrations of 95 g/L, the switch to xylose utilization was severely stalled such that efficient xylose fermentation could not occur.\n",
      "Ethanol concentrations around 50 g/L fully repressed enzyme induction although xylose transport into the cells was observed to be occurring.\n",
      "Increasing degrees of repression were documented between 15 and 45 g/L ethanol.\n",
      "Repitched cell populations grown on xylose resulted in faster fermentation rates, particularly on xylose but also on glucose, and eliminated diauxic lag and stalling during mixed sugar conversion by P. tannophilus or S. stipitis, despite ethanol accumulations in the 60 or 70 g/L range, respectively.\n",
      "78 ../prodcurator/6721273_curator.json\n",
      "79 ../prodcurator/17188753_curator.json\n",
      "80 ../prodcurator/25301663_curator.json\n",
      "Rice seedlings grown under 50 muM cadmium alone or in combination with 5 muM methyl jasmonate were investigated for Cd-induced oxidative injury at 3, 7 and 10/ds of treatment.\n",
      "81 ../prodcurator/22877686_curator.json\n",
      "Based on the low production of H(2)S, acetic acid and foam, ethanol resistance, growth in presence of high concentrations of potassium metabisulphite (KMBS) and CuSO(4) and at low temperatures, 14 strains were selected and used as starter to ferment grape must at 13  C and 17  C in presence of 100 mg/L of KMBS.\n",
      "82 ../prodcurator/10861414_curator.json\n",
      "A gradual improvement of the inoculum build-up procedure, culture medium composition, and process parameters ultimately led to a theoretically maximal L-sorbose productivity (200 g /L of L-sorbose from 200 g /L of D-sorbitol in 28 h of fermentation) using a Gluconobacter oxydans mutant strain that was selected under conditions of substrate inhibition.\n",
      "83 ../prodcurator/9792721_curator.json\n",
      "84 ../prodcurator/8936555_curator.json\n",
      "85 ../prodcurator/25604674_curator.json\n",
      "Under the above conditions, the extraction yields of echinacoside and acteoside (the main components of phenylethanoid glycosides) reached 5.35 and 6.22 mg/g dry material weight, respectively.\n",
      "The contents of echinacoside and acteoside in the extracts reached 27.56 and 30.23 mg/g, respectively, which were 2.46- and 2.58-fold higher than the amounts obtained in ultrasound-assisted extraction.\n",
      "86 ../prodcurator/2168329_curator.json\n",
      "87 ../prodcurator/20926158_curator.json\n",
      "Short-term Al(3+) toxicity was induced by a 60-min exposure to a pH-buffered solution (pH 4.5) containing 2mM CaSO(4) and 50muM AlCl(3).\n",
      "88 ../prodcurator/29131629_curator.json\n",
      "The maximum theanderose concentration ranged from 122.4 to 130.4 g/L , was obtained after only 1 h and at a moderate temperature (37  C), leading to high productivity (109.7-130.4 g/L/h-1) and yield (up to 37.3%) values.\n",
      "89 ../prodcurator/8624354_curator.json\n",
      "90 ../prodcurator/180235_curator.json\n",
      "In contrast to previous findings with Saccharomyces carlsbergensis, cytochrome c oxidase and another mitochondrial enzyme, succinate dehydrogenase, were completely sedimentable by zonal centrifugation in sucrose gradients in the presence of either 2 mM-MgCl2 or 0-4 mM-EDTA.\n",
      "91 ../prodcurator/1705962_curator.json\n",
      "Glycine (5-10 mM) completely eliminated the depolarizing response of the C-cell, whereas the hyperpolarizing component was unaffected.\n",
      "In contrast, gamma-aminobutyric acid (GABA; 5-10 mM) had no  effect on either component.\n",
      "The C-cell light response was modified in two stages by cis-2,3-piperidine dicarboxylic acid (cis-PDA; 0.5-5 mM): first the depolarizing response disappeared; then the membrane potential hyperpolarized concomitant with a large reduction or elimination of the hyperpolarizing light response.\n",
      "92 ../prodcurator/1104625_curator.json\n",
      "A stable Escherichia coli IF-2-fMet-tRNAfMet complex is formed upon incubation of IF-2 (prokaryotic initiation factor) with fMet-tRNAfMet is the presence of 50 mM Tris-HCl, pH 7.0, 100 mM NH4Cl, and 7 mM 2-mercaptoethanol.\n",
      "Complex formation does dot require GTP, is unstable in the presence of 5 mM Mg2+, and is specific for fMet-tRNAfMet.\n",
      "93 ../prodcurator/16678321_curator.json\n",
      "Although Cu is phytotoxic at Cu(2+) activities as low as 1-2 microM, the effect of Cu(2+) on the nodulation of legumes has received  attention.\n",
      "The nodulation process was more sensitive to increasing Cu(2+) activities than both shoot and root growth; whilst a Cu(2+) activity of 1.0 microM corresponded to a 10% reduction in the relative yield of the shoots and roots, a Cu(2+) activity of 0.2 microM corresponded to a 10% reduction in nodulation.\n",
      "This reduction in nodulation with increasing Cu(2+) activity was associated with an inhibition of root hair formation in treatments containing > or =0.77 microM Cu(2+), rather than to a reduction in the size of the Rhizobium population.\n",
      "94 ../prodcurator/8349665_curator.json\n",
      "95 ../prodcurator/4934522_curator.json\n",
      "96 ../prodcurator/25272315_curator.json\n",
      "97 ../prodcurator/15885621_curator.json\n",
      "The inactivation of these genes in SBS550MG increased the succinate yield from glucose to about 1.6 mol/mol with an average anaerobic productivity rate of 10 mM/h (approximately 0.64 mM/h-OD600).\n",
      "98 ../prodcurator/1090597_curator.json\n",
      "Cells of Escherichia coli strain B develop large intracellular vacuoles and exhibit an abnormally long lag phase when inoculated into a defined medium to glucose and salts containing 3 times 10-6 M Cd2+.\n",
      "99 ../prodcurator/8061051_curator.json\n",
      "We have shown that ethanol, propanol and butanol (at 0.5-2%) and salts of butyric and propionic acids (at 8-40 mM) all cause a major reduction in heat-shock protein (hsp) synthesis when present in the growth medium of Drosophila cultured cells (Kc and SL2) subjected to either increased temperature or chemical stressors.\n",
      "100 ../prodcurator/3503645_curator.json\n",
      "101 ../prodcurator/17039388_curator.json\n",
      "In the present investigation, the previous ultraviolet irradiated mutant strain of Aspergillus oryzae UV-7 was further improved in terms of 3,4 dihydroxy phenyl L-alanine (L-DOPA) activity after chemical mutagenesis through 1-methyl 3-nitro 1-nitroso guanidine (MNNG = 250-1500 microg/ml) treatment (0-30 min).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Relatively better production of L-DOPA (> 0.60 mg/ml) was obtained when 2.0% (w/v) glucose was used as a carbon source in the mycelium production medium and the tyrosinase activity increased constitutively (1.08 mg/ml), which resulted in a greater production of L-DOPA .\n",
      "The maximum conversion of L-tyrosine to L-DOPA (0.428 mg/ml) was achieved 60 min after the biochemical reaction.\n",
      "Best results of L-DOPA biosynthesis were observed when the concentration of illite was 3.5 x 10(-6) M (1.686 mg/ml L-DOPA produced with 1.525 mg/ml consumption of L-tyrosine).\n",
      "The comparison of kinetic parameters showed the ability of mutant to yield L-DOPA (i.e., Yp/x 7.360 +/- 0.04 mg/mg).\n",
      "102 ../prodcurator/15135955_curator.json\n",
      "Buffers and water containing 10 mM NaCl became bactericidal against all three bacteria upon PEF treatment, and the bactericidal effect could be neutralized by thiosulfate, suggesting that chlorine and/or hypochlorite had been formed.\n",
      "In the Ames mutagenicity test using His- S .\n",
      "103 ../prodcurator/23796348_curator.json\n",
      "Employing Lactobacillus brevis ADH, it was possible to achieve the synthesis of enantiopure (S)-fluorohydrins at a 0.5 M substrate concentration.\n",
      "104 ../prodcurator/16716244_curator.json\n",
      "105 ../prodcurator/26460058_curator.json\n",
      "500 mg/L without external ethanol supplementation.\n",
      "106 ../prodcurator/28783799_curator.json\n",
      "107 ../prodcurator/19851774_curator.json\n",
      "The major changes upon a 150 mM NaCl exposure in potato leaves occurred in the photosystem apparatus and Calvin cycle: many transcripts coding for proteins belonging to photosystems I and II and chlorophyll synthesis were repressed.\n",
      "108 ../prodcurator/18595087_curator.json\n",
      "Continuous sucrose feeding (5 g/L/h) to the immobilized-cell culture in the airlift bioreactor increased production of enzymatic activity by about 107% compared with ordinary batch operation of this reactor.\n",
      "109 ../prodcurator/1369308_curator.json\n",
      "110 ../prodcurator/15039991_curator.json\n",
      "111 ../prodcurator/14734837_curator.json\n",
      "In the stirred-tank fermentor, the highest concentration of the carotenoid pigment (92.0 mg/L) was obtained at an aeration rate of 1.5 vvm and agitation speed of 60 rpm.\n",
      "Maximum beta-carotene concentration (360.2 mg/L) was obtained in culture grown in molasses solution containing 5% (w/v) sugar supplemented with linoleic acid (37.59 g/L), kerosene (39.11 g/L), and antioxidant (1.0 g/L).\n",
      "112 ../prodcurator/26818577_curator.json\n",
      "89.5% COD, 90.6% TN and 91.2% TP removals were achieved as well as the highest biomass production of 2655mg/L and ALA yield of 7.40mg/g-biomass under the optimal HRT of 60h and OLR of 2.48 g/L/d.\n",
      "113 ../prodcurator/7476556_curator.json\n",
      "Growth rates, viability, and protein concentrations of E. coli grown in the absence of glucose were unaffected by the addition of 100 microM CCCP .\n",
      "The culture of E. coli appeared to be in a syncopic state hovering between life and death when it was exposed to CCCP in the presence of more than 30 mM glucose.\n",
      "114 ../prodcurator/11337972_curator.json\n",
      "115 ../prodcurator/16968109_curator.json\n",
      "In planta, 82% of group 1 strains and 25% of group 2 strains produced high  (100-1500 microg/g) of moniliformin.\n",
      "All group 2 strains from more northern states produced  or no moniliformin (0-5 microg/g).\n",
      "116 ../prodcurator/22113557_curator.json\n",
      "117 ../prodcurator/650511_curator.json\n",
      "118 ../prodcurator/845139_curator.json\n",
      "119 ../prodcurator/1387860_curator.json\n",
      "The strain synthesized the compound in culture media containing  2 microM added FeCl3.\n",
      "120 ../prodcurator/11601540_curator.json\n",
      "Acetic acid production was 0.67 g/g cellulose substrate utilized in cellulose mineral salt (CMS) medium.\n",
      "121 ../prodcurator/16721882_curator.json\n",
      "122 ../prodcurator/6804443_curator.json\n",
      "When cells were allowed to become inactive by suspending them in the dark in Mg2+-free buffer, full activity could be restored with a few minutes by adding 20 mM Mg2+ and illuminating the cells.\n",
      "123 ../prodcurator/16194150_curator.json\n",
      "124 ../prodcurator/7591051_curator.json\n",
      "Residual contaminating iron, 0.45 microM, was sufficient to support iron-limited growth.\n",
      "Typical iron-replete cultures metabolized 3.3 microM iron.\n",
      "125 ../prodcurator/31332677_curator.json\n",
      "This mixed consortium was used as inoculum for biogas production using different concentrations of sugarcane bagasse (SCB) (from 1.58 to 4.42 g/L) and yeast extract (YE) (from 0.58 to 3.42 g/L) according to a composite central design.\n",
      "The maximum ethanol production (20.11 mg/L) was observed using 2.0 and 3.0 g/L  of YE and SCB, respectively (C6).\n",
      "Likewise, the highest hydrogen production (0.60 mmol/L) was observed using 3.0 and 4.0 g/L  of YE and SCB, respectively (C1).\n",
      "Methane was also observed, reaching the maximum production (1.44 mmol/L) using 1.0 and 4.0 g/L  of YE and SCB, respectively (C2).\n",
      "126 ../prodcurator/7835650_curator.json\n",
      "127 ../prodcurator/2007885_curator.json\n",
      "Honeybees have approximately 2.5 pmol/eye of retinal associated with visual pigments, but larger amounts (4-12 pmol/eye) of both retinal and retinol bound to soluble proteins.\n",
      "128 ../prodcurator/17120457_curator.json\n",
      "At various hydraulic retention times, ranging from 0.7 to 5/ds, excellent removal of chemical oxygen demand (COD), biochemical oxygen demand (BOD), total suspended solids (TSS), ammonia-nitrogen (NH4-N), and total Kjeldahl nitrogen was achieved with final effluent BOD/TSS/NH4N concentrations of  15/15/1 mg/L .\n",
      "129 ../prodcurator/10869302_curator.json\n",
      "130 ../prodcurator/10457057_curator.json\n",
      "131 ../prodcurator/17496338_curator.json\n",
      "For this purpose, we used media containing glucose (4.3-6.5 g/L), xylose (60.1-92.1 g/L), or arabinose (5.9-9.2 g/L), or binary or ternary mixtures of them in either the presence or absence of typical inhibitors of acidic hydrolysates, such as furfural (1.0-5.0 g/L), hydroxymethylfurfural (0.01- 0.30 g/L), acetic acid (0.5-3.0 g/L), and vanillin (0.5-3.0 g/L).\n",
      "132 ../prodcurator/26981892_curator.json\n",
      "We combined a number of strategies to create a base strain producing 11.4 mg/L of the precursor geraniol.\n",
      "133 ../prodcurator/26018342_curator.json\n",
      "The XYL1-overexpressed mutant produced xylitol from D-xylose using glycerol as a co-substrate for cell growth and NAD (P) H regeneration: 100 g/L D-xylose was completely converted into xylitol when at least 20 g/L glycerol was used as a co-substrate.\n",
      "XYL1 overexpressed mutant produced xylitol with a volumetric productivity of 0.83 g/L/h, and a xylitol yield of 98 % xylose.\n",
      "134 ../prodcurator/23604959_curator.json\n",
      "135 ../prodcurator/29727857_curator.json\n",
      "Residual sugar  were  2.6 mg/g while the shell contents and ethanol were in the range of 11-13.4% and 4.8-7 mg/g, respectively.\n",
      "Beans fermented in the presence of LAB contained higher  of lactic acid (0.6-1.2 mg/g) whereas higher concentrations of acetic acid (1.8-2.2 mg/g) were detected in beans inoculated with AAB.\n",
      "136 ../prodcurator/26063803_curator.json\n",
      "137 ../prodcurator/20221789_curator.json\n",
      "The highest protein yield was obtained by: transforming HbII into the BLi5 cells, inducing and supplementing the culture during the mid-log phase with 1 mM IPTG, 30 microg/mL hemin chloride and 1% glucose, and decreasing the temperature to 30 degrees C after induction.\n",
      "138 ../prodcurator/6645996_curator.json\n",
      "A temperature shock, a change in the pH of the medium for conservation within the range of 4.0 to 10.0, and an increase of NaCl concentration up to 5 M do not inactivate Escherichia coli phages T3, T4 and phi X174.\n",
      "139 ../prodcurator/3118395_curator.json\n",
      "140 ../prodcurator/24031679_curator.json\n",
      "A culture of Saccharomyces cerevisiae M30 entrapped in loofa-reinforced alginate was used for continuous ethanol fermentation in a packed-bed reactor with initial sugar concentrations of 200-248 g/L .\n",
      "Maximum ethanol productivity of 11.5 g/L/h was obtained at an ethanol concentration of 57.4 g/L, an initial sugar concentration of 220 g/L and a dilution rate (D) of 0.2/h.\n",
      "However, a maximum ethanol concentration of 82.1 g/L (productivity of 9.0 g/L/h) was obtained at a D of 0.11/h.\n",
      "141 ../prodcurator/15302885_curator.json\n",
      "142 ../prodcurator/15281930_curator.json\n",
      "For example, 9C can be cultured over a wide range of temperature, pH and salt concentration, and showed high H2 production (up to 67.7 nmol H2/h 10(10) cell(-1)).\n",
      "143 ../prodcurator/24756896_curator.json\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A two-liquid phase fed-batch fermentation with glucose as the sole carbon and energy source resulted in the formation of 700 mg /L (S)-limonene.\n",
      "The use of glycerol as a carbon source resulted in a prolonged growth and production phase (specific activities of >=50 mU g(cdw) (-1) ) leading to a final (S)-limonene concentration of 2,700 mg /L .\n",
      "144 ../prodcurator/30125151_curator.json\n",
      "145 ../prodcurator/17222854_curator.json\n",
      "The method consisted of a two-step gradient (0% B for 2 min and from 0 to 50% B in 10 min) being mobile phase A a 2 0mM borate buffer (pH 9) and mobile phase B a 20 mM borate buffer (pH 9) containing 1M sodium chloride.\n",
      "146 ../prodcurator/20045990_curator.json\n",
      "147 ../prodcurator/2369580_curator.json\n",
      "148 ../prodcurator/8407961_curator.json\n",
      "149 ../prodcurator/18811000_curator.json\n",
      "Application of 100 microM aspterric acid (AA), a pollen growth inhibitor, with different concentrations of indole-3-acetic acid (IAA) results in the recovery of normal pollen development of Arabidopsis thaliana.\n",
      "Treatment with 100 microM AA plus 5 mM IAA  induced the normal seed production.\n",
      "Treatment with 100 microM N-1-naphthylphthalamic acid (NPA), a polar auxin transport inhibitor, did not reduce the pollen growth but inhibited seed production.\n",
      "100 microM NPA plus 5 mM IAA did not induce any seed production.\n",
      "The endogenous level of IAA in stems and leaves of A. thaliana treated with 100 microM AA was similar to that of the untreated control.\n",
      "In contrast to AA treatment, the IAA level by the treatment with 100 microM NPA was about twice as much as that of the untreated control.\n",
      "150 ../prodcurator/28887657_curator.json\n",
      "Two novel aphid-resistance loci, Rag6 and Rag3c, from Glycine soja 85-32, were previously detected in a 10.5-cM interval on chromosome 8 and a 7.5-cM interval on chromosome 16, respectively.\n",
      "151 ../prodcurator/9172777_curator.json\n",
      "152 ../prodcurator/9772191_curator.json\n",
      "153 ../prodcurator/16095939_curator.json\n",
      "In a fed-batch culture, the DG1(pSPD5) strain was able to produce 1,3-propanediol at a higher concentration (1104 mM) and productivity than the natural producer C. butyricum VPI 3266.\n",
      "Furthermore, this strain was also successfully used for very long term continuous production of 1,3-propanediol at high volumetric productivity (3 g/L/h) and titer (788 mM).\n",
      "154 ../prodcurator/12529516_curator.json\n",
      "155 ../prodcurator/28284309_curator.json\n",
      "Finally, further overexpression of OCrtZ and OBKTM in YastD-01 resulted in accumulation of 8.10mg/g  (47.18mg/l) of (3S, 3'S)-astaxanthin in shake-flask cultures.\n",
      "156 ../prodcurator/17302316_curator.json\n",
      "In a plot with 100% of CW percolate irrigation, the maximum Zn, Mn and Cu concentrations of 5.0, 12.3 and 2.5 mg/kg, respectively, were detected in the percolate-fed soil, whereas no accumulation of heavy metals in the plant tissues (i.e. leaves, stems and flowers) of the sunflower were detected.\n",
      "157 ../prodcurator/2076050_curator.json\n",
      "158 ../prodcurator/805816_curator.json\n",
      "159 ../prodcurator/6435534_curator.json\n",
      "When irradiated with AzIN at concentrations approaching IN saturation for the IN----Trp activity (0.1 mM), in the absence of serine, the enzyme was increasingly inactivated (up to 70-80%) concomitant with the progressive binding of a net of 2 mol AzIN per alpha beta equivalent.\n",
      "Little or no cooperativity in the binding of the 2 mol AzIN was observed.\n",
      "160 ../prodcurator/2934360_curator.json\n",
      "161 ../prodcurator/20499163_curator.json\n",
      "The minimum inhibitory concentration of Cd(2+) against C. tropicalis was 2,500 mg /L .\n",
      "The yeast also showed tolerance toward Zn(2+) (1,400 mg /L), Ni(2+) (1,000 mg /L), Hg(2+) (1,400 mg /L), Cu(2+) (1,000 mg /L), Cr(6+) (1,200 mg /L), and Pb(2+) (1,000 mg /L).\n",
      "The metal processing ability of the isolate was determined in a medium containing 100 mg /L of Cd(2+).\n",
      "162 ../prodcurator/9363636_curator.json\n",
      "Some protection against this inactivation is provided by high concentrations of dithiothreitol (20 mM), suggesting that the inactivation may be due to chemical oxidation.\n",
      "163 ../prodcurator/7217785_curator.json\n",
      "164 ../prodcurator/27189644_curator.json\n",
      "165 ../prodcurator/16049735_curator.json\n",
      "166 ../prodcurator/10103262_curator.json\n",
      "Thermal inactivation kinetic analysis of ThMA indicated that it was stabilized in the presence of 10 mM EDTA.\n",
      "167 ../prodcurator/22000965_curator.json\n",
      "Under optimum hydrolysis conditions, 0.685 g/g of reducing sugar was produced per gram of pretreated biomass.\n",
      "The fermentation of the hydrolyzate using Saccharomyces cerevisae produced 11.365 g/L of bioethanol with an efficiency of about 50%.\n",
      "168 ../prodcurator/24186657_curator.json\n",
      "During chilling, glucose-6-phosphate and fructose-6-phosphate accumulated up to 16 mM (assuming they are restricted to the cytoplasm).\n",
      "169 ../prodcurator/10892859_curator.json\n",
      "170 ../prodcurator/2834333_curator.json\n",
      "171 ../prodcurator/15764651_curator.json\n",
      "The studied system, composed of approximately 300,000 atoms, included one copy of the protein, a patch of a DPPC lipid bilayer, and a 1 M water solution of KCl.\n",
      "172 ../prodcurator/19855940_curator.json\n",
      "The hydrolysates could readily be fermented without addition of nutrients and with average fermentation yields of 0.43 +/- 0.02 g/g based on consumed mannose and glucose.\n",
      "Employing SSF, final ethanol concentrations of 70 g/kg was achieved in 216 h, corresponding to an ethanol yield of 70% mol/mol of theoretical or 200 g ethanol/kg PKC.\n",
      "173 ../prodcurator/30558416_curator.json\n",
      "The best engineered strain using high-throughput fermentation produced squalene from glucose at 5.4 +- 0.3 mg/g   and 105.3 +- 3.0 mg/L, which was a 5.2-fold increase over the parental strain.\n",
      "In addition, flask cultivation of C. glutamicum overexpressing the dxs and idi genes with squalene synthase gene and repressing the idsA gene resulted in production of squalene at 5.8 +- 0.4 mg/g  and 82.8 +- 6.2 mg/L, which was a 3.4-fold increase over the parental strain.\n",
      "174 ../prodcurator/20120465_curator.json\n",
      "175 ../prodcurator/27746096_curator.json\n",
      "Fed-batch fermentations of PALFK with low- and medium-density (OD600 of 0.4 and 9.0, respectively) inocula produced 69.2 and 78.4 g/L of homo-SA with yields of 1.56 and 1.64 mol/mol glucose equivalent and overall volumetric SA productivities of 2.50 and 6.02 g/L/h, respectively, using sucrose and glycerol as dual carbon sources.\n",
      "The SA productivity could be further increased to 38.6 g/L/h by employing a membrane cell recycle bioreactor system.\n",
      "176 ../prodcurator/22818948_curator.json\n",
      "Nitrogen waste was primarily in the form of dissolved inorganic nitrogen (TDIN) and was removed by 98.2 +- 2.2% under ambient loadings of 109-383 mumol/L .\n",
      "At peak loadings of up to 8185 +- 590 mumol/L (equivalent to 600 mmol N m(-2) d(-1)), the filter beds removed between 30 and 58% (250 mmol N m(-2) d(-1)) of influent TDIN .\n",
      "Influent dissolved inorganic phosphorus  ranged from 34 to 90 mumol/L, with 36-89% reduction under routine operations.\n",
      "Dissolved organic nitrogen (DON) loadings were lower (11-144 mumol/L), and between 23 and 69% of influent DON was removed during routine operation, with no significant removal of DON under high TDIN loading.\n",
      "Over the 88-day study, cumulative nitrogen removal was 1.28 mol m(-2), of which 1.09 mol m(-2) was retained in plant tissue, with plant uptake ranging from 2.4 to 27.0 mmol N g(-1) dry weight d(-1).\n",
      "177 ../prodcurator/24821237_curator.json\n",
      "178 ../prodcurator/16740394_curator.json\n",
      "Using synthetic growth medium we were able to express in L. tarentolae and purify to homogeneity (15)N-valine labeled Enchanced Green Fluorescent Protein (EGFP) with the final yield of 5.7 mg/L of suspension culture.\n",
      "179 ../prodcurator/10358127_curator.json\n",
      "Contractions were abolished by tetrodotoxin (1.6 microM).\n",
      "Apamin (300 nM) disrupted evoked peristaltic waves and  increased the rate-of-rise of the LM and CM contractions anal to a stimulus, and the LM oral to a stimulus.\n",
      "Nomega-nitro-L-arginine (100 microM), a nitric oxide synthesis inhibitor, had no overall significant effect on the characteristics of the LM and CM contractions, although on occasion an enhancement in their peak amplitude was noted.\n",
      "180 ../prodcurator/6481646_curator.json\n",
      "181 ../prodcurator/18347834_curator.json\n",
      "182 ../prodcurator/7829985_curator.json\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Insulin-like growth factor (IGF)-I (10(-9)M-10(-7)M) was found to be mitogenic for cultured cells from all three zones as shown by increased incorporation of [3H]thymidine into DNA.\n",
      "183 ../prodcurator/9325277_curator.json\n",
      "TrwD is probably a peripheral outer membrane protein because it could be solubilized by increasing salt concentration to 0.5 M NaCl in the lysis buffer.\n",
      "184 ../prodcurator/9439593_curator.json\n",
      "185 ../prodcurator/20883441_curator.json\n",
      "Using an A. thaliana microarray chip, we determined changes in the expression of approximately 2 800 genes between A. stelleri plants treated with 0.2 M mannitol versus mock-treated plants.\n",
      "japonica seedlings treated with 0.2 M mannitol, 0.2 M sorbitol, and 0.2 M NaCl.\n",
      "186 ../prodcurator/9398079_curator.json\n",
      "Herbicide 14C-terbutryn binding measurements, also with thylakoids, show that the QB niche of the mutant is  modified, at least 7-8 fold increased terbutryn dissociation constant is shown (220 nM in the mutant versus 29 nM in the wild type); the PS II sensitivity to bicarbonate-reversible formate inhibition is reduced by 5 fold in the mutant, although the formate/bicarbonate binding site still exists in the mutant.\n",
      "187 ../prodcurator/105901_curator.json\n",
      "188 ../prodcurator/3279015_curator.json\n",
      "189 ../prodcurator/19434429_curator.json\n",
      "According to the mathematic models, the optimal semi-continuous condition for total yield of polysaccharide was NaNO3 3.5 g/L, renewal rate 27%, renewal period 2.91/ds.\n",
      "The optimal condition for polysaccharide output rate was NaNO3 0.5 g/L, renewal rate 5%, renewal period 7/ds.\n",
      "The maximum output rate of polysaccharide was 68.64 mg/L per/d, which was 2.02 times higher than previous reported data.\n",
      "190 ../prodcurator/9770642_curator.json\n",
      "Using 14C-isotopes, the cycling of NAH between lens and ocular fluids in a simple support medium consisting of NaCl (0.9%), Ca2+ (4 mEq/L) and D-glucose (5 mM) at pH 7.4 has previously been observed.\n",
      "With each NAH molecule recycled up to 10 times/d, and a power factor of 33, there could be 330 mmol of water transported/mmol of NAH each/d.\n",
      "With typical NAH concentrations in brains of poikilothermic vertebrates of 5-10 mmol/kg, there is the potential for up to 3.3 mol (60 mL) of water to be removed each/d/kg of brain, a value that represents about 8% of total brain water content.\n",
      "191 ../prodcurator/19054232_curator.json\n",
      "192 ../prodcurator/6588537_curator.json\n",
      "193 ../prodcurator/10788790_curator.json\n",
      "Although  2 mM ALA increased porphyrin synthesis, ALA  above 3 mM inhibited the synthesis.\n",
      "The porphyrin lesion produced in low intensity light (300 lx) was considerably reduced by 100 microM M874 glycolipid, although the reduction in intense light (3,000 lx) was restricted to a lower level.\n",
      "194 ../prodcurator/27475463_curator.json\n",
      "195 ../prodcurator/3956741_curator.json\n",
      "196 ../prodcurator/15138858_curator.json\n",
      "197 ../prodcurator/26549611_curator.json\n",
      "198 ../prodcurator/3007727_curator.json\n",
      "199 ../prodcurator/18767207_curator.json\n",
      "200 ../prodcurator/17141752_curator.json\n",
      "The following CPA mixtures were evaluated: (1) 0.45M trehalose and 0, 0.64, 1.02 and 1.53 M dimethyl sulfoxide (Me(2)SO); (2) 0.2M glucose and 2M Me(2)SO and (3) 1.31 M propylene glycol (PG).\n",
      "This study showed that total motility was best retained when spermatozoa were cryopreserved in 0.45 M trehalose and 0, 0.64, 1.02 or 1.53 M Me(2)SO and frozen using slow to medium-rapid cooling rates (2.1-5.2 degrees Cmin(-1)).\n",
      "Rapid freezing through direct plunging of spermatozoa into liquid nitrogen resulted in the lowest overall retention of motility regardless of the CPA additive; however, CPA mixture also influenced retention of motility, with 0.2M glucose in 2M Me(2)SO and 1.31 M PG retaining the lowest  of motility for the CPAs evaluated.\n",
      "201 ../prodcurator/11722901_curator.json\n",
      "202 ../prodcurator/26556681_curator.json\n",
      "Aspergillus flavus growth and AFB1 production were completely inhibited by 0.80 mmol/L of cinnamaldehyde and 2.80 mmol/L of citral.\n",
      "However, at lower concentration, cinnamaldehyde (0.40 mmol/L), eugenol (0.80 mmol/L), and citral (0.56 mmol/L)  reduced AFB1 production with inhibition rate of 68.9%, 95.4%, and 41.8%, respectively, while no effect on fungal growth.\n",
      "Real-time PCR showed that the expressions of aflR, aflT, aflD, aflM, and aflP were down-regulated by cinnamaldehyde (0.40 mmol/L), eugenol (0.80 mmol/L), and citral (0.56 mmol/L).\n",
      "With 0.80 mmol/L of eugenol, aflP was highly down-regulated (average of 2061-folds), followed by aflM, aflR, aflD, and aflT with average of 138-, 15-, 5.2-, and 4.8-folds reduction, respectively.\n",
      "With 0.56 mmol/L of citral, aflT was completely inhibited, followed by aflM, aflP, aflR, and aflD with average of 257-, 29-, 3.5-, and 2.5-folds reduction, respectively.\n",
      "203 ../prodcurator/27734668_curator.json\n",
      "The d-tagatose production at 500 g/L d-galactose by the host was 1.4-fold higher than that by Escherichia coli expressing d-GaI .\n",
      "PICG cells, which produced 165 g/L d-tagatose after 3 h, with a conversion of 55% (w/w) and a productivity of 55 g/L/h, showed  higher d-tagatose productivity than that reported for other cells.\n",
      "204 ../prodcurator/29430908_curator.json\n",
      "205 ../prodcurator/17058077_curator.json\n",
      "206 ../prodcurator/18355019_curator.json\n",
      "The presence of 100 mM L-arabinose does not influence the response of the reported mutants to D-arabinose, and the mutants are not induced by other sugars tested (D-xylose, D-fucose, D-lyxose).\n",
      "Co-expression of the FucP transporter in E. coli enabled induction by D-arabinose in the 0.1 mM range.\n",
      "207 ../prodcurator/16687162_curator.json\n",
      "Based on DO production, median effective concentration (EC50) varies from 1.73 mg/L (1-octanol) to 8,040 mg/L (2-propanol).\n",
      "208 ../prodcurator/21603950_curator.json\n",
      "A maximum ethanol concentration of 29.7 g/L was achieved after 120 h of fermentation; however, after 90 h of fermentation, the ethanol concentration was only slightly lower at 29.1 g/L with a yield of 0.39 g ethanol per gram of sugar consumed.\n",
      "Using the same air flow rate and adjusting the agitation rate resulted in lower ethanol yields of 0.25 g/g at 50 rpm and 0.30 g/g at 300 rpm.\n",
      "209 ../prodcurator/1444399_curator.json\n",
      "HV content in the polymer was increased from 17 to 24 mol% by adding 10 to 40 mM valerate to glucose medium, but HV insertion into the polymer occurred at a fixed rate.\n",
      "The synthesis of HV in these cultures also occurred at a fixed rate (2.3 to 2.8 mol% h-1), while the maximum PHA production rate was 1.1 g/L/h-1.\n",
      "Up to 45 mol% C source, but the PHA produced amounted to  1 g/L .\n",
      "The combination of 30 mM valerate as a sole C source and 0.5 mM 4-pentenoate increased the HV content in the polymer to 52 mol%.(ABSTRACT TRUNCATED AT 250 WORDS)\n",
      "210 ../prodcurator/10886_curator.json\n",
      "17900 and a turnover number of 4s-1 (50mM-triethanolamine/400mM-KCl, pH 7.2, 25 degrees C) with dihydrofolate and NADPH as substrates.\n",
      "Michaelis constants for a variety of substrates have been measured by using a new fluorimetric assay (0.36 muM-dihydrofolate; 0.78 muM-NADPH), and binding constants determined by using the quenching of protein fluorescence (dihydrofolate, 2.25 X 10(6)M-1; NADPH, greater than 10(8)M-1).\n",
      "211 ../prodcurator/16080719_curator.json\n",
      "Culture medium containing 25 g /L of activated carbon gave an optimal surfactin yield of 3600 mg /L, which was approximately 36-fold higher than that obtained from carrier-free liquid culture.\n",
      "Concentration of limiting carbon substrate (glucose) is also an important factor affecting the production of surfactin, as an initial glucose concentration of 40 g /L resulted in optimal surfactin production.\n",
      "The purified surfactin reduced the surface tension of water from 72 to 27 mN m(-1) with a critical micelle concentration of ca. 10 mg /L .\n",
      "The surfactin product also attained an emulsion index of 70% for kerosene and diesel at a low concentration of 100 and 600 mg /L, respectively.\n",
      "212 ../prodcurator/15027622_curator.json\n",
      "213 ../prodcurator/23292549_curator.json\n",
      "When cultivated in the presence of 50 mM aspartate, the biomass of cells increased 65.8 % compared with the control (without aspartate addition).\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chem_word</th>\n",
       "      <th>prod_word</th>\n",
       "      <th>chem_word_idx</th>\n",
       "      <th>prod_word_idx</th>\n",
       "      <th>sentence</th>\n",
       "      <th>tokens</th>\n",
       "      <th>between_tokens</th>\n",
       "      <th>chems</th>\n",
       "      <th>prods</th>\n",
       "      <th>chem_pos_word</th>\n",
       "      <th>label</th>\n",
       "      <th>pubmedid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alpha-D-glucose</td>\n",
       "      <td>0.2mM</td>\n",
       "      <td>[5, 10]</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.</td>\n",
       "      <td>[As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]</td>\n",
       "      <td>[added]</td>\n",
       "      <td>[27*#*28, 5*#*10, 12*#*17, 30*#*36]</td>\n",
       "      <td>[29*#*30, 3*#*4, 11*#*12]</td>\n",
       "      <td>{'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}</td>\n",
       "      <td>-1</td>\n",
       "      <td>3955571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>alpha-D-glucose</td>\n",
       "      <td>25mM</td>\n",
       "      <td>[5, 10]</td>\n",
       "      <td>[29, 30]</td>\n",
       "      <td>As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.</td>\n",
       "      <td>[As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]</td>\n",
       "      <td>[(, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from]</td>\n",
       "      <td>[27*#*28, 5*#*10, 12*#*17, 30*#*36]</td>\n",
       "      <td>[29*#*30, 3*#*4, 11*#*12]</td>\n",
       "      <td>{'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}</td>\n",
       "      <td>-1</td>\n",
       "      <td>3955571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>alpha-D-xylose</td>\n",
       "      <td>0.2mM</td>\n",
       "      <td>[12, 17]</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.</td>\n",
       "      <td>[As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]</td>\n",
       "      <td>[added, alpha, -, D, -, glucose, (, 0.4mM]</td>\n",
       "      <td>[27*#*28, 5*#*10, 12*#*17, 30*#*36]</td>\n",
       "      <td>[29*#*30, 3*#*4, 11*#*12]</td>\n",
       "      <td>{'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}</td>\n",
       "      <td>-1</td>\n",
       "      <td>3955571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>alpha-D-xylose</td>\n",
       "      <td>25mM</td>\n",
       "      <td>[12, 17]</td>\n",
       "      <td>[29, 30]</td>\n",
       "      <td>As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.</td>\n",
       "      <td>[As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]</td>\n",
       "      <td>[),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from]</td>\n",
       "      <td>[27*#*28, 5*#*10, 12*#*17, 30*#*36]</td>\n",
       "      <td>[29*#*30, 3*#*4, 11*#*12]</td>\n",
       "      <td>{'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}</td>\n",
       "      <td>-1</td>\n",
       "      <td>3955571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fluoride</td>\n",
       "      <td>0.2mM</td>\n",
       "      <td>[27, 28]</td>\n",
       "      <td>[3, 4]</td>\n",
       "      <td>As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.</td>\n",
       "      <td>[As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]</td>\n",
       "      <td>[added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of]</td>\n",
       "      <td>[27*#*28, 5*#*10, 12*#*17, 30*#*36]</td>\n",
       "      <td>[29*#*30, 3*#*4, 11*#*12]</td>\n",
       "      <td>{'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}</td>\n",
       "      <td>-1</td>\n",
       "      <td>3955571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chem_word prod_word chem_word_idx prod_word_idx  \\\n",
       "0  alpha-D-glucose  0.2mM     [5, 10]       [3, 4]         \n",
       "1  alpha-D-glucose  25mM      [5, 10]       [29, 30]       \n",
       "2  alpha-D-xylose   0.2mM     [12, 17]      [3, 4]         \n",
       "3  alpha-D-xylose   25mM      [12, 17]      [29, 30]       \n",
       "4  fluoride         0.2mM     [27, 28]      [3, 4]         \n",
       "\n",
       "                                                                                                                                                                      sentence  \\\n",
       "0  As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.   \n",
       "1  As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.   \n",
       "2  As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.   \n",
       "3  As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.   \n",
       "4  As  as 0.2mM added alpha-D-glucose (0.4mM alpha-D-xylose)  increased the rate of enzymically catalyzed release of fluoride from 25mM beta-D-glucosyl fluoride at 0 degrees.   \n",
       "\n",
       "                                                                                                                                                                                                                                  tokens  \\\n",
       "0  [As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]   \n",
       "1  [As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]   \n",
       "2  [As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]   \n",
       "3  [As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]   \n",
       "4  [As,  , as, 0.2mM, added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from, 25mM, beta, -, D, -, glucosyl, fluoride, at, 0, degrees, .]   \n",
       "\n",
       "                                                                                                                            between_tokens  \\\n",
       "0  [added]                                                                                                                                   \n",
       "1  [(, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from]                   \n",
       "2  [added, alpha, -, D, -, glucose, (, 0.4mM]                                                                                                \n",
       "3  [),  , increased, the, rate, of, enzymically, catalyzed, release, of, fluoride, from]                                                     \n",
       "4  [added, alpha, -, D, -, glucose, (, 0.4mM, alpha, -, D, -, xylose, ),  , increased, the, rate, of, enzymically, catalyzed, release, of]   \n",
       "\n",
       "                                 chems                      prods  \\\n",
       "0  [27*#*28, 5*#*10, 12*#*17, 30*#*36]  [29*#*30, 3*#*4, 11*#*12]   \n",
       "1  [27*#*28, 5*#*10, 12*#*17, 30*#*36]  [29*#*30, 3*#*4, 11*#*12]   \n",
       "2  [27*#*28, 5*#*10, 12*#*17, 30*#*36]  [29*#*30, 3*#*4, 11*#*12]   \n",
       "3  [27*#*28, 5*#*10, 12*#*17, 30*#*36]  [29*#*30, 3*#*4, 11*#*12]   \n",
       "4  [27*#*28, 5*#*10, 12*#*17, 30*#*36]  [29*#*30, 3*#*4, 11*#*12]   \n",
       "\n",
       "                                                                                                              chem_pos_word  \\\n",
       "0  {'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}   \n",
       "1  {'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}   \n",
       "2  {'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}   \n",
       "3  {'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}   \n",
       "4  {'5*#*10': 'alpha-D-glucose', '12*#*17': 'alpha-D-xylose', '27*#*28': 'fluoride', '30*#*36': 'beta-D-glucosyl fluoride'}   \n",
       "\n",
       "  label pubmedid  \n",
       "0  -1    3955571  \n",
       "1  -1    3955571  \n",
       "2  -1    3955571  \n",
       "3  -1    3955571  \n",
       "4  -1    3955571  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle,json,os\n",
    "import pandas as pd\n",
    "def is_decimal(num):\n",
    "    dnumre = re.compile(r\"^[0-9]+(\\.[0-9]{1,2})?$\")\n",
    "    result = dnumre.search(num)\n",
    "    return bool(result)\n",
    "\n",
    "def add_single_unit(text):\n",
    "    nlp = nlp_change()\n",
    "    doc = nlp(text)\n",
    "    units_dict = defaultdict(list)\n",
    "    prod_start_end = {}\n",
    "    for ent in doc.ents:\n",
    "        #print(dir(ent))\n",
    "        #print(ent.text, ent.start_char, ent.end_char, ent.label_,ent.start,ent.end)\n",
    "        prod_start_end[ent.start] = ent.end \n",
    "\n",
    "    for tok in doc:\n",
    "        #print(tok.text,\"-->\", tok.pos_,\"-->\",tok.ent_type_,\"-->\",tok.i)\n",
    "        if tok.i<1:\n",
    "            continue\n",
    "        if tok.ent_type_ in [\"TITER\",\"YIELDS\"]:\n",
    "            if not tok.nbor(-1).ent_type_:\n",
    "                prod_end = prod_start_end[tok.i]\n",
    "                unite_key = tok.i\n",
    "                units_dict[unite_key].append(prod_end)\n",
    "            else:\n",
    "                try:\n",
    "                    units_dict[unite_key].append(tok.i)\n",
    "                except:\n",
    "                    continue\n",
    "    unit_text_dict = {}\n",
    "    for digital,unit_list in units_dict.items():\n",
    "        if len(unit_list) >1:\n",
    "            #print(\"#\"*20,\"unit_list\",unit_list)\n",
    "            unit_text = doc[unit_list[1]:unit_list[0]]\n",
    "            unit_text_dict[unit_list[1]] = unit_text\n",
    "\n",
    "        #print(\"unit_text\",unit_text)\n",
    "    new_doc = \"\"\n",
    "    init_num = 0\n",
    "    unit_num = len(unit_text_dict.keys())\n",
    "    for tok in doc:\n",
    "        tok_index = tok.i\n",
    "        \n",
    "        if tok.pos_ == \"NUM\" :\n",
    "            tok_text = tok.text\n",
    "            if not is_decimal(tok_text):\n",
    "                continue\n",
    "            try:\n",
    "                tok_position = tok.i\n",
    "                nbor_1 = tok.nbor(1)\n",
    "                nbor_1_text = nbor_1.text\n",
    "                if nbor_1_text not in [\",\",\"and\"]:\n",
    "                    continue                               \n",
    "                nbor_2 = tok.nbor(2)\n",
    "                nbor_2_pos = nbor_2.i + 2 - 1 # 2:neigbor distance; 1: token position(-1)\n",
    "                if nbor_2_pos in unit_text_dict.keys(): \n",
    "                    unit = unit_text_dict[nbor_2_pos]\n",
    "                    unit_text = unit.text\n",
    "                    tok_prod = tok.text\n",
    "                    new_doc = new_doc + doc[init_num:tok_position+1].text + \" \" + unit_text + \" \"\n",
    "                    #print(new_doc)\n",
    "                    init_num = tok_position + 1\n",
    "                    del unit_text_dict[nbor_2_pos]\n",
    "                    unit_text_keys_len = len(unit_text_dict.keys())\n",
    "            except:\n",
    "                pass\n",
    "    new_doc = new_doc + doc[init_num:].text\n",
    "    return new_doc\n",
    "def add_multipe_units(text):\n",
    "    \n",
    "    if \"respectively\" not in text:\n",
    "        return text\n",
    "    #print(\"&*\"*200,text)        \n",
    "    text = re.sub('(?<=\\d), and (?=\\d)', ' and ', text, re.S)\n",
    "    for i in range(4):\n",
    "        newtext = add_single_unit(text)\n",
    "        if not newtext:\n",
    "            break\n",
    "        if newtext != text:       \n",
    "            text = newtext        \n",
    "        else:\n",
    "            break            \n",
    "    return text\n",
    "\n",
    "\n",
    "\n",
    "def newtext_token(text):\n",
    "    new_text = add_multipe_units(text)\n",
    "    nlp = nlp_change()\n",
    "    doc = nlp(new_text)\n",
    "    for tok in doc:\n",
    "        print(tok.text,\"-->\", tok.pos_,\"-->\",tok.ent_type_,\"-->\",tok.i)\n",
    "    for ent in doc.ents:\n",
    "        #print(dir(ent))\n",
    "        print(ent.text, ent.start_char, ent.end_char, ent.label_,ent.start,ent.end)\n",
    "def get_text_between(chem_word_idx,prod_word_idx,tokens):\n",
    "    \"\"\"\n",
    "    Returns the text between the two person mentions in the sentence for a candidate\n",
    "    \"\"\"\n",
    "\n",
    "    chem_start = chem_word_idx[0]\n",
    "    chem_end = chem_word_idx[1]\n",
    "    prod_start = prod_word_idx[0]\n",
    "    prod_end = prod_word_idx[1]  \n",
    "    if chem_end < prod_end:\n",
    "        start = chem_end\n",
    "        end = prod_start\n",
    "    else:\n",
    "        start = prod_end\n",
    "        end = chem_start\n",
    "\n",
    "    tokens_between =tokens[start:end]\n",
    "    return tokens_between\n",
    "def adv_delete_process(text):\n",
    "    text = text.replace('significantly','')\n",
    "    text = text.replace('levels','')\n",
    "    text = text.replace('obvious','')\n",
    "    text = text.replace('little','')\n",
    "    text = text.replace('substantially','')\n",
    "    text = text.replace('less than','')\n",
    "    return text\n",
    "def abstract_process(text):\n",
    "    text = adv_delete_process(text)\n",
    "    text = text.replace(' g L-1',' g/L ')\n",
    "    #text = text.replace('L-1',' /L')\n",
    "    text = text.replace(' L-1','/L')\n",
    "    text = text.replace('L(-1)', '/L')\n",
    "    text = text.replace(' l(-)(1)','/L')\n",
    "    text = text.replace(' l(-1)','/L')\n",
    "    text = text.replace('L(org) (-1)','/L')\n",
    "    text = text.replace('/L. ', '/L . ')\n",
    "    text = text.replace(' d. ', ' d . ')\n",
    "    text = text.replace(' . g','/g')\n",
    "\n",
    "    text = text.replace('g/liter','g/L')\n",
    "    text = text.replace('g liter-1','g/L')\n",
    "    text = text.replace('grams per liter','g/L')\n",
    "    text = text.replace('by-product','byproduct')\n",
    "    text = text.replace('g (dry weight)','g')\n",
    "    text = text.replace('each (','')\n",
    "    text = text.replace('(L h)','L/h')\n",
    "    text = text.replace('/L h','/L/h')\n",
    "    text = text.replace('  h-1','/h')\n",
    "    text = text.replace(' h(-1)','/h')\n",
    "    text = text.replace('/100 g','/g')\n",
    "    text = text.replace(' day-1','/d')\n",
    "    text = text.replace(' day','/d')\n",
    "    text = text.replace('dry cell weight','DCW')\n",
    "    text = text.replace(\"DCW\",'')\n",
    "    text = text.replace('()','')\n",
    "    text = re.sub('(?<=g)L-1(?=d)', '/L ', text, re.S)\n",
    "    text = re.sub('(?<=\\d)g(?=/)', ' g', text, re.S)\n",
    "    text = re.sub('(?<=/.)g day', 'g/d', text, re.S)\n",
    "    text = re.sub('(?<=\\d)mol(?=/)',' mol',text,re.S)\n",
    "    text = re.sub('(?<=[a-z])A\\. (?=[A-Z])', 'A . ', text, re.S)\n",
    "    text = re.sub(' h\\. (?=[A-Z])', ' h . ', text, re.S)\n",
    "    text = re.sub('(?<=[A-Z])\\. (?=[A-Z])', ' . ', text, re.S)\n",
    "    text = re.sub('% molar yield', '% mol/mol', text, re.S)\n",
    "    return text\n",
    "def get_pubmedid_from_path(path_text):\n",
    "    path,filename = os.path.split(path_text)\n",
    "    pubmedid_raw,file_type = os.path.splitext(filename)\n",
    "    pubmed_id = pubmedid_raw[:-8]\n",
    "    return pubmed_id\n",
    "\n",
    "path_df = \"../ProcessedData/all_synbio20200716.pkl\"\n",
    "with open(path_df, 'rb') as data:\n",
    "    df = pickle.load(data)\n",
    "pmid_list = list(df['PMID'])\n",
    "abs_title_error_lemma = ['body', 'diet', 'KD', 'Km','mean','degrade','degradation','infect','infected','ingest', 'infuse', 'Ki', 'LC50', 'serum', 'Kd', 'K(M', 'K(m','K(i', 'rabbit', 'dose', 'perfuse', 'reperfusion', 'IC50', 'inject', 'dosage']\n",
    "abs_title_error_lemma_set = set(abs_title_error_lemma)\n",
    "\n",
    "sp = spacy.load(\"en_core_web_sm\") \n",
    "\n",
    "\n",
    "error_list = set()\n",
    "head_column = [\"chem_word\",\"prod_word\",\"chem_word_idx\",'prod_word_idx','sentence','tokens',\\\n",
    "               'between_tokens','chems','prods','chem_pos_word',\"label\",'pubmedid']\n",
    "result =pd.DataFrame(columns=head_column)\n",
    "for i,filename in enumerate(glob(\"../prodcurator/*curator.json\")):\n",
    "    print(i,filename)\n",
    "    pubmed_id = get_pubmedid_from_path(filename)\n",
    "    pmid_index = pmid_list.index(pubmed_id)\n",
    "    #print(pmid_index)\n",
    "    abs_title = df.loc[pmid_index].Title_Abstract\n",
    "    abs_title_token = sp(abs_title)\n",
    "    abs_title_lemma = [word.lemma_ for word in abs_title_token]\n",
    "    abs_title_lemma_set = set(abs_title_lemma)\n",
    "    if len(abs_title_lemma_set.intersection(abs_title_error_lemma_set))>0:\n",
    "        continue\n",
    "    annotator_text = json.load(open(filename))\n",
    "    for key,sentence_curator in annotator_text.items():\n",
    "        sub_chem_prod_list = []\n",
    "        sentence_text = sentence_curator['sentece_text']\n",
    "        sentence_text = abstract_process(sentence_text)\n",
    "        new_text = add_multipe_units(sentence_text)\n",
    "        nlp = nlp_change()\n",
    "        doc = nlp(new_text)\n",
    "        sentence_tokens = [tok.text for tok in doc]\n",
    "        sentence_chems_set = set()\n",
    "        sentence_prods_set = set()\n",
    "        chem_prod_list = []\n",
    "        chem_pos_word = {}\n",
    "        \n",
    "        for sub_key, text_info in sentence_curator.items():\n",
    "            if sub_key == 'sentece_text':\n",
    "                continue\n",
    "            sub_chem_prod_dict = {} \n",
    "            if sub_key == \"PRO\":\n",
    "                curator_label_dig = 1\n",
    "            else:\n",
    "                curator_label_dig = -1            \n",
    "            for positive_unit in text_info:\n",
    "                \n",
    "                chem = positive_unit['chem']\n",
    "                chempos = positive_unit['chempos']\n",
    "                prod_lable = positive_unit['prod_lable']\n",
    "                if prod_lable == 0:\n",
    "                    prod_lable_dig = -1\n",
    "                else:\n",
    "                    prod_lable_dig = 1\n",
    "                curator_label = curator_label_dig * prod_lable_dig\n",
    "                chempos_text = \"*#*\".join(chempos)\n",
    "                sentence_chems_set.add(chempos_text)\n",
    "                new_chempos = [int(chemposdig) for chemposdig in chempos]\n",
    "                old_chem = doc[new_chempos[0]:new_chempos[1]].text\n",
    "                chem_pos_word[chempos_text] = old_chem\n",
    "                if old_chem != chem:\n",
    "                    error_list.add(filename)\n",
    "                    #print(\"#\"*20,doc[new_chempos[0]:new_chempos[1]],chem)\n",
    "                   \n",
    "                \n",
    "                \n",
    "                prod = positive_unit['prod']\n",
    "                prodpos = positive_unit['prodpos']\n",
    "                prodpos_text = \"*#*\".join(prodpos)\n",
    "                sentence_prods_set.add(prodpos_text)\n",
    "                new_prodpos = [int(prodposdig) for prodposdig in prodpos]\n",
    "                old_prod = doc[new_prodpos[0]:new_prodpos[1]].text\n",
    "                if old_prod != prod:\n",
    "                    error_list.add(filename)\n",
    "                    \n",
    "                \n",
    "                text_between = get_text_between(new_chempos,new_prodpos,sentence_tokens)\n",
    "                sub_chem_prod_dict = {'chem':chem,\\\n",
    "                                      'chempos':new_chempos,\\\n",
    "                                      'prod':prod,\\\n",
    "                                      'prodpos':new_prodpos,\\\n",
    "                                      'between_tokens':text_between,\\\n",
    "                                      'Label':curator_label}\n",
    "                chem_prod_list.append(sub_chem_prod_dict)\n",
    "\n",
    "            \n",
    "            \n",
    "        sentence_chems = list(sentence_chems_set)\n",
    "        sentence_prods = list(sentence_prods_set)\n",
    "        for chem_prod in chem_prod_list:\n",
    "            chem_word = chem_prod['chem']\n",
    "            \n",
    "            chem_word_idx = chem_prod[\"chempos\"]\n",
    "            \n",
    "            prod_word = chem_prod['prod']\n",
    "            prod_word_idx = chem_prod['prodpos']\n",
    "            #print(\"#\"*20,type(prod_word_idx),prod_word_idx)\n",
    "            label = chem_prod[\"Label\"]\n",
    "            between_tokens = chem_prod['between_tokens']\n",
    "\n",
    "            temp = pd.DataFrame({\n",
    "                                 \"chem_word\":[chem_word],\\\n",
    "                                 \"chem_word_idx\":[chem_word_idx],\\\n",
    "                                 \"prod_word\":[prod_word],\\\n",
    "                                 'prod_word_idx':[prod_word_idx],\\\n",
    "                                 'sentence':[sentence_text],\\\n",
    "                                 'tokens':[sentence_tokens],\\\n",
    "                                 'between_tokens': [between_tokens],\\\n",
    "                                 'chems':[sentence_chems],\\\n",
    "                                 'prods':[sentence_prods],\\\n",
    "                                 'chem_pos_word':[chem_pos_word],\\\n",
    "                                 \"label\":[label],\\\n",
    "                                 \"pubmedid\":[pubmed_id]\n",
    "                                 })            \n",
    "            result = pd.concat([result, temp], ignore_index=True)\n",
    "                                         \n",
    "        print(sentence_text)\n",
    "pd.set_option(\"display.max_colwidth\", 0)\n",
    "result = result.reindex(columns=head_column) \n",
    "result.reset_index(drop=True)\n",
    "\n",
    "#result.to_pickle(\"../ProcessedData/procon_corpus.pkl\")\n",
    "result.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
